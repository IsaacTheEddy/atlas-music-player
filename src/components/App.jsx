import Footer from "./Footer";
import MusicPlayer from "./MusicPlayer";
import { PlayList } from "./PlayList";
import { Skelton } from "./Skeleton/Skeleton";
import { useState } from "react";

function App() {
  const [a, setA] = useState(false);
  return (
    <>
      {a === false ? (
        <Skelton />
      ) : (
        <div className="flex h-full min-h-screen flex-col bg-KindaGray dark:bg-DarkChocalate dark:text-white">
          <div className="mx-auto flex flex-col rounded-2xl border-8 border-double border-Mocha md:flex-row">
            <div className="min-h-80% flex h-fit flex-col justify-between rounded-es-lg border-Lightpink p-4 md:border-r-4 dark:border-Mocha">
              <MusicPlayer />
            </div>
            <div className="min-h-80% flex h-fit flex-col justify-between p-4">
              <PlayList />
            </div>
          </div>
          <Footer />
        </div>
      )}
    </>
  );
}

export default App;
