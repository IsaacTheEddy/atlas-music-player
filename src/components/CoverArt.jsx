import React from "react";
import cover from "../assets/placeholder.svg";

export function CoverArt() {
  return (
    <img
      className="size-100 flex-1 rounded-2xl"
      src={cover}
      alt="placeholder"
    />
  );
}
