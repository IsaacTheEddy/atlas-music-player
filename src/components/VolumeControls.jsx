import React, { useState } from "react";
import volume from "../assets/volume-2.svg";
import muted from "../assets/volume-x.svg";

function handleClick(info) {
  return alert(info);
}

export function VolumeControl() {
  const [range, setRange] = useState(5);
  const [mute, setMute] = useState(false);

  const handleRangeChange = (event) => {
    setRange(event.target.value);
  };

  const handleMute = (mute) => {
    if (mute === false) {
      alert(`Mute is ${mute}`);
      setMute(true);
    } else if (mute === true) {
      alert(`Mute is  ${mute}`);
      setMute(false);
    }
  };

  return (
    <div className="mt-2.5 flex flex-row justify-between">
      <button onClick={() => handleMute(mute)}>
        {mute === false ? (
          <img src={muted} alt="" className="size-15" />
        ) : (
          <img src={volume} alt="" className="size-15" />
        )}
      </button>
      <input
        className="ml-0 w-90/100 justify-self-center accent-stone-500/50 ease-in-out"
        type="range"
        name=""
        value={range}
        min="0"
        max="10"
        onChange={handleRangeChange}
      />
      <p>{mute}</p>
    </div>
  );
}
