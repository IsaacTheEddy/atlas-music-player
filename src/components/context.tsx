import React, { createContext, useContext } from "react";
import { DataType } from "./App";

interface CurrentSongContextType {
  playlist?: [];
  currentSong: DataType;
  setCurrentSong?: React.Dispatch<React.SetStateAction<DataType>>;
  currentSongIndex?: number;
  setCurrentSongIndex?: React.Dispatch<React.SetStateAction<number>>;
  setPlaylist?: React.Dispatch<React.SetStateAction<DataType[]>>;
  isPlaying?: boolean;
  setIsPlaying?: React.Dispatch<React.SetStateAction<boolean>>;
  shuffle?: boolean;
  setShuffle?: React.Dispatch<React.SetStateAction<boolean>>;
  mute?: boolean;
  setMute?: React.Dispatch<React.SetStateAction<boolean>>;
  volume?: number;
  setVolume?: React.Dispatch<React.SetStateAction<number>>;
  playback?: 0.5 | 1.0 | 2.0;
  setPlayback?: React.Dispatch<React.SetStateAction<0.5 | 1.0 | 2.0>>;
  duration?: number | undefined;
  setDuration?: React.Dispatch<React.SetStateAction<number | undefined>>;
  loading?: boolean;
  setLoading?: React.Dispatch<React.SetStateAction<boolean>>;
  playNextSong?: () => void;
  playPreviousSong?: () => void;
}

export const currentSongContext = createContext<
  CurrentSongContextType | undefined
>(undefined);

export function useCurrentSongContext() {
  const currentsong = useContext(currentSongContext);

  if (!currentsong) {
    throw new Error("Use current song context correctly scrub");
  }
  return currentsong;
}
