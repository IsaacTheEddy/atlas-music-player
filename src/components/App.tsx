import Footer from "./Footer";
import MusicPlayer from "./MusicPlayer";
import PlayListItem from "./PlayListItem";
import { currentSongContext } from "./context";
import { Skelton } from "./Skeleton/Skeleton";
import React, { useEffect, useState } from "react";

export interface DataType {
  id: string;
  title: string;
  artist: string;
  genre: string;
  duration: number;
  cover: string;
  song: string;
  isPlaying: boolean;
}
const PAGE_SIZE = 1;

export const App: React.FC = () => {
  const [playlist, setPlaylist] = useState<DataType[]>([]);
  const [currentSong, setCurrentSong] = useState<DataType>({
    id: "a6pkp78whsyqdvpb5dxn64ss",
    title: "Fading Shadows",
    artist: "The Emberlight",
    genre: "Alternative Rock",
    duration: 194,
    cover: "https://utfs.io/f/E9fJnaKtTy1bV09oPkISuh6fWpNsTRlAk1Qj9yqnVzCi32BL",
    song: "https://utfs.io/f/E9fJnaKtTy1ba1N97yFrtlodEaJI0m4wGY9KgyPiUvCf8hMp",
    isPlaying: false,
  });
  const [currentSongIndex, setCurrentSongIndex] = useState<number>(0);

  const [isPlaying, setIsPlaying] = useState<boolean>(false);
  const [mute, setMute] = useState<boolean>(false);
  const [volume, setVolume] = useState<number>(5);
  const [playback, setPlayback] = useState<0.5 | 1 | 2>(1);
  const [shuffle, setShuffle] = useState<boolean>(true);
  const [loading, setLoading] = useState<boolean>(true);
  const [duration, setDuration] = useState<number | undefined>(undefined);

  useEffect(() => {
    const api = async () => {
      try {
        const response = await fetch("api/v1/playlist");
        if (!response.ok) {
          throw new Error("Error there");
        }
        const jsonData: DataType[] = await response.json();
        setPlaylist(jsonData);
        if (jsonData.length > 0) {
          setCurrentSong(jsonData[0]);
          setCurrentSongIndex(0);
        }
      } catch (e: any) {
        throw new Error("Error here");
      } finally {
      }
    };
    api();
  }, []);

  const playNextSong = () => {
    if (playlist.length === 0) return;
    const nextIndex = currentSongIndex + 1;
    if (
      nextIndex < playlist.length &&
      currentSong.id !== playlist[nextIndex].id
    ) {
      setLoading(true);
      setCurrentSongIndex(nextIndex);
      setCurrentSong(playlist[nextIndex]);
      setIsPlaying(true);
    } else if (nextIndex === 10) {
      setLoading(true);
      setCurrentSongIndex(0);
      setCurrentSong(playlist[0]);
      setIsPlaying(true);
    } else {
      setLoading(true);
      setCurrentSongIndex(nextIndex + 1);
      setCurrentSong(playlist[nextIndex + 1]);
      setIsPlaying(true);
    }
  };

  const playPreviousSong = () => {
    if (playlist.length === 0) return;

    const prevIndex = currentSongIndex - 1;
    if (prevIndex > -1 && currentSong.id !== playlist[prevIndex].id) {
      setLoading(true);
      setCurrentSongIndex(prevIndex);
      setCurrentSong(playlist[prevIndex]);
      setIsPlaying(true);
    } else if (prevIndex === -1) {
      setLoading(true);
      setCurrentSongIndex(9);
      setCurrentSong(playlist[9]);
      setIsPlaying(true);
    } else {
      setLoading(true);
      setCurrentSongIndex(prevIndex - 1);
      setCurrentSong(playlist[prevIndex - 1]);
      setIsPlaying(true);
    }
  };

  return (
    <>
      <currentSongContext.Provider
        value={{
          playlist,
          currentSong,
          setCurrentSong,
          currentSongIndex,
          setCurrentSongIndex,
          isPlaying,
          setIsPlaying,
          volume,
          setVolume,
          mute,
          setMute,
          playback,
          setPlayback,
          playNextSong,
          playPreviousSong,
          shuffle,
          setShuffle,
          duration,
          setDuration,
          loading,
          setLoading,
        }}
      >
        <div className="flex h-full min-h-screen flex-col bg-KindaGray dark:bg-DarkChocalate dark:text-white">
          <div className="mx-auto mt-6 flex flex-col rounded-2xl border-8 border-double border-Mocha md:flex-row">
            <div className="min-h-80% flex h-fit flex-col justify-between rounded-es-lg border-Lightpink p-4 md:border-r-4 dark:border-Mocha">
              <MusicPlayer />
            </div>
            <div className="min-h-80% flex h-fit flex-col justify-between p-4">
              <PlayListItem />
            </div>
          </div>
          <Footer />
        </div>
      </currentSongContext.Provider>
    </>
  );
};

export default App;
