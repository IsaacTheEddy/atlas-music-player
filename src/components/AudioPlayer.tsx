import React, { useEffect, useRef, useState } from "react";
import { useCurrentSongContext } from "./context";

type Props = {};

export const AudioPlayer = () => {
  const ref = useRef<HTMLAudioElement>(null);
  const song = useRef<HTMLAudioElement>(null);
  const {
    currentSong,
    setCurrentSong,
    isPlaying,
    volume,
    mute,
    playback,
    playNextSong,
    duration,
    setDuration,
  } = useCurrentSongContext();
  useEffect(() => {
    if (!ref.current) {
      return;
    }
    if (isPlaying) {
      ref.current.play();
    } else {
      ref.current.pause();
    }
  }, [isPlaying]);

  useEffect(() => {
    if (!ref.current) {
      return;
    }
    if (!volume) {
      return;
    }
    if (mute === true) {
      ref.current.volume = 0;
    } else {
      ref.current.volume = volume / 10;
    }
  }, [volume, mute]);

  useEffect(() => {
    if (!ref.current) {
      return;
    }
    ref.current.playbackRate = playback;
  }, [playback]);

  useEffect(() => {
    if (!ref.current) {
      return;
    }
    ref.current.src = currentSong.song;
  }, [currentSong]);

  return (
    <audio ref={ref} onEnded={() => playNextSong()} autoPlay>
      <source type="audio/mpeg"></source>
    </audio>
  );
};
