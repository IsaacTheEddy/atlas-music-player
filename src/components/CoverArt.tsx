import { useCurrentSongContext } from "./context";

export function CoverArt() {
  const c = useCurrentSongContext();
  return (
    <>
      <img
        className="size-100 flex-1 rounded-2xl"
        src={c.currentSong.cover}
        alt="placeholder"
      />
    </>
  );
}
