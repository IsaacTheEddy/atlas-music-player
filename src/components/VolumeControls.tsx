import React, { useState } from "react";
import { useCurrentSongContext } from "./context";

export function VolumeControl() {
  const { volume, setVolume } = useCurrentSongContext();
  const { mute, setMute } = useCurrentSongContext();

  const handleRangeChange = (event: any) => {
    setVolume(event.target.value);
  };

  const handleMute = () => {
    if (mute === false) {
      setMute(true);
    } else if (mute === true) {
      setMute(false);
    }
  };

  return (
    <div className="mt-2.5 flex flex-row justify-between">
      <button onClick={() => handleMute()}>
        {mute === true ? (
          <svg
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            stroke-linecap="round"
            stroke-linejoin="round"
            className="size-7"
          >
            <path d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z" />
            <line x1="22" x2="16" y1="9" y2="15" />
            <line x1="16" x2="22" y1="9" y2="15" />
          </svg>
        ) : (
          <svg
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            stroke-linecap="round"
            stroke-linejoin="round"
            className="size-7"
          >
            <path d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z" />
            <path d="M16 9a5 5 0 0 1 0 6" />
            <path d="M19.364 18.364a9 9 0 0 0 0-12.728" />
          </svg>
        )}
      </button>
      {mute === true ? (
        <>
          <input
            disabled
            className="ml-0 w-90/100 justify-self-center accent-Mocha ease-in-out dark:accent-Lightsoftie/75"
            type="range"
            name=""
            value={volume}
            min="0"
            max="10"
            onChange={handleRangeChange}
          />
          <p>{volume}</p>
        </>
      ) : (
        <>
          <input
            className="ml-0 w-90/100 justify-self-center accent-Mocha ease-in-out dark:accent-Lightsoftie/75"
            type="range"
            name=""
            value={volume}
            min="0"
            max="10"
            onChange={handleRangeChange}
          />
          <p>{volume}</p>
        </>
      )}
    </div>
  );
}
