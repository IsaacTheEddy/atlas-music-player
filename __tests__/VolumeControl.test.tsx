import { render } from "@testing-library/react";
import { test, expect } from "vitest";
import { VolumeControl } from "../src/components/VolumeControls";
import {
  currentSongContext,
  useCurrentSongContext,
} from "../src/components/context";

test("If mute is enabled(mute = true)", () => {
  let { container } = render(
    <currentSongContext.Provider
      value={{
        playlist: [],
        currentSong: {
          id: "a6pkp78whsyqdvpb5dxn64ss",
          title: "Fading Shadows",
          artist: "The Emberlight",
          genre: "Alternative Rock",
          duration: 194,
          cover:
            "https://utfs.io/f/E9fJnaKtTy1bV09oPkISuh6fWpNsTRlAk1Qj9yqnVzCi32BL",
          song: "https://utfs.io/f/E9fJnaKtTy1ba1N97yFrtlodEaJI0m4wGY9KgyPiUvCf8hMp",
          isPlaying: false,
        },
        mute: true,
      }}
    >
      <VolumeControl />
    </currentSongContext.Provider>,
  );

  expect(container).toMatchSnapshot();
});
test("If mute is disabled (mute = false)", () => {
  let { container } = render(
    <currentSongContext.Provider
      value={{
        playlist: [],
        currentSong: {
          id: "a6pkp78whsyqdvpb5dxn64ss",
          title: "Fading Shadows",
          artist: "The Emberlight",
          genre: "Alternative Rock",
          duration: 194,
          cover:
            "https://utfs.io/f/E9fJnaKtTy1bV09oPkISuh6fWpNsTRlAk1Qj9yqnVzCi32BL",
          song: "https://utfs.io/f/E9fJnaKtTy1ba1N97yFrtlodEaJI0m4wGY9KgyPiUvCf8hMp",
          isPlaying: false,
        },
        mute: true,
      }}
    >
      <VolumeControl />
    </currentSongContext.Provider>,
  );

  expect(container).toMatchSnapshot();
});
