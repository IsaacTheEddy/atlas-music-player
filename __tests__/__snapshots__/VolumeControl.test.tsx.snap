// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`If mute is disabled (mute = false) 1`] = `
<div>
  <div
    class="mt-2.5 flex flex-row justify-between"
  >
    <button>
      <svg
        class="size-7"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
        />
        <line
          x1="22"
          x2="16"
          y1="9"
          y2="15"
        />
        <line
          x1="16"
          x2="22"
          y1="9"
          y2="15"
        />
      </svg>
    </button>
    <input
      class="ml-0 w-90/100 justify-self-center accent-Mocha ease-in-out dark:accent-Lightsoftie/75"
      disabled=""
      max="10"
      min="0"
      name=""
      type="range"
    />
    <p />
  </div>
</div>
`;

exports[`If mute is enabled(mute = true) 1`] = `
<div>
  <div
    class="mt-2.5 flex flex-row justify-between"
  >
    <button>
      <svg
        class="size-7"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
        />
        <line
          x1="22"
          x2="16"
          y1="9"
          y2="15"
        />
        <line
          x1="16"
          x2="22"
          y1="9"
          y2="15"
        />
      </svg>
    </button>
    <input
      class="ml-0 w-90/100 justify-self-center accent-Mocha ease-in-out dark:accent-Lightsoftie/75"
      disabled=""
      max="10"
      min="0"
      name=""
      type="range"
    />
    <p />
  </div>
</div>
`;
